create_shards.c

```diff
@@ -91,12 +103,69 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 		ereport(ERROR, (errmsg("could not find column: %s", partitionColumnName)));
 	}
 
-	/* we only support hash partitioning method for now */
-	if (partitionMethod != HASH_PARTITION_TYPE)
+	partitionColumnVar = ColumnNameToColumn(distributedTableId, partitionColumnName);
+	partitionColumnTypeId = partitionColumnVar->vartype;
+
+	if (partitionMethod == HASH_PARTITION_TYPE)
+	{
+		partitionColumnOpClassId = GetDefaultOpClass(partitionColumnTypeId, HASH_AM_OID);
+	}
+	else
 	{
+		/* we only support hash partitioning right now */
 		ereport(ERROR, (errmsg("unsupported partition method: %c", partitionMethod)));
 	}
 
+	/*
+	 * If data type of the partition column does not have a default operator class, we
+	 * should prevent distribution of the table on this column. It is because there are
+	 * no operators/procedures that can manipulate the value of the column.
+	 */
+	if (partitionColumnOpClassId == InvalidOid)
+	{
+		ereport(ERROR, (errcode(ERRCODE_WRONG_OBJECT_TYPE),
+						errmsg("cannot distribute relation: \"%s\"", tableName),
+						errdetail("Column \"%s\" does not have a default operator class"
+								  " with the provided partition method.",
+								  partitionColumnName),
+						errhint("Try another column as the distribution column.")));
+	}
+
+	/* depending on the partition type, check for the existence of support procedure */
+	if (partitionMethod == HASH_PARTITION_TYPE)
+	{
+		hashSupportRoutine = GetSupportRoutine(partitionColumnTypeId, HASH_AM_OID,
+											   HASHPROC);
+		if (hashSupportRoutine == InvalidOid)
+		{
+			ereport(ERROR, (errcode(ERRCODE_WRONG_OBJECT_TYPE),
```

Comment:
Again, `ERRCODE_WRONG_OBJECT_TYPE` is inappropriate. `arrayfuncs.c` has a _ton_ of examples of errors to use when a specific function lookup fails. Something like:

```c
ereport(ERROR,
		(errcode(ERRCODE_UNDEFINED_FUNCTION),
		 errmsg("could not identify a comparison function for type %s",
		 		format_type_be(element_type))
		 errdetail("Partition column types must have a comparison function defined to use range partitioning.")));
```